sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -702,3 +702,46 @@ def test_single_node_trees(Est):
 def test_custom_loss(Est, loss, X, y):
     est = Est(loss=loss, max_iter=20)
     est.fit(X, y)
+
+
+@pytest.mark.parametrize(
+    'scoring, validation_fraction, early_stopping, n_iter_no_change, tol', [
+        ('neg_mean_squared_error', .1, True, 5, 1e-7),  # use scorer
+        ('neg_mean_squared_error', None, True, 5, 1e-1),  # use scorer on train
+        (None, .1, True, 5, 1e-7),  # same with default scorer
+        (None, None, True, 5, 1e-1),
+        ('loss', .1, True, 5, 1e-7),  # use loss
+        ('loss', None, True, 5, 1e-1),  # use loss on training data
+        (None, None, False, 5, None),  # no early stopping
+        ])
+def test_staged_predict(scoring, validation_fraction,
+                        early_stopping, n_iter_no_change, tol):
+
+    from sklearn.utils._testing import assert_raises
+    from sklearn.utils._testing import assert_array_almost_equal
+    # Test whether staged decision function eventually gives
+    # the same prediction.
+    X, y = make_regression(n_samples=1200, random_state=0)
+
+    X_train, y_train = X[:200], y[:200]
+    X_test = X[200:]
+    clf = HistGradientBoostingRegressor(
+        scoring=scoring,
+        tol=tol,
+        early_stopping=early_stopping,
+        validation_fraction=validation_fraction,
+        n_iter_no_change=n_iter_no_change,
+        random_state=0
+    )
+    # test raise ValueError if not fitted
+    assert_raises(ValueError, lambda X: np.fromiter(
+        clf.staged_predict(X), dtype=np.float64), X_test)
+
+    clf.fit(X_train, y_train)
+    y_pred = clf.predict(X_test)
+
+    # test if prediction for last stage equals ``predict``
```

Comment:
that's good

We can be even more thorough and  check something like:

```py
    for i, y_pred in enumerate(reg.staged_predict(X_test), 1):
        aux = GBRegressor(max_iter=i)
		assert_allclose(y_pred, aux.predict(X_test))
```

Id suggest setting max_iter of `reg` to e.g. 10 for the test to be fast

We should also make sure that the number of iterations in staged_predict is equal to reg.n_iter_

and we should also add tests for the limit case of max_iter = 1

Reply:
Thanks for the review!
I will change the style and terminology for more consistency.

About the tests, should we add separated tests like:

- One for testing final predictions
- Another for testing each stage
- Another for testing limit case of max_iter = 1
- Another for testing the number of iterations in staged_predict is equal to reg.n_iter_

P.S. I can also add staged_predict for classification!